Fix verify signature on encrypted email from Apple Mail

jeffhsta opened the pull request:
The leap.mail has a step that decrypt an email and at the same time it should verify signature.
It works when using Thunderbird to send an email, but when using Apple Mail that step always return "could not verify signature" because of the way that Apple Mail encrypt and sign the email.

Issue: https://github.com/pixelated/pixelated-user-agent/issues/605

kalikaneko commented:
can you rebase this PR against latest develop?
DEBUG mode is not there anymore, and pycryptopp is not a dependency.

jeffhsta commented:
@kalikaneko sure, it's my bad, I will fix this.

Commit:
[bug] verify signature encrypted email from Apple Mail

 Fix verify signature on encrypted email from Apple Mail, adding a step
 to verify signature after decrypt the email because the keymananger
 could not verify signature when decrypting it

jeffhsta commented:
@kalikaneko is that right now?

kalikaneko commented:
@jeffhsta yes, thanks!

jeffhsta commented:
@kalikaneko sure, I will get the payload of both to compare between Thunderbird and Apple Mail.

Basically what Thunderbird does is encrypt and sign the email like "gpg -se <mail>", so when leap.mail decrypts the email, it will verify the signature at the same time.

Apple Mail first sign the email and then encrypt it like "gpg -s <email> | gpg -e <signed_email>", so when leap.mail decrypt the email, it get a signed plain text email, because of that I added one step more after decrypt an email.

I will add a break point to show emails from Thunderbird and Apple Mail here to make it more clear to understand

jeffhsta commented:
@kalikaneko Here is the result that I got debugging the leap.mail

Added a ipdb break point on line 529 on my commit version.

@kalikaneko any questions you can ping me on IRC as well

** Apple Mail:

signkey: ```InvalidSignature(u'Failed to verify signature with key D08CDC5EDA9C5095',)```

str(decrdata):
```
From nobody Thu Feb 25 20:26:50 2016
Return-Path: <thais@domain.com>
X-Original-To: me@pixelated.org
Delivered-To: 8823007cd15b6b2b72b15517da7630a9@deliver.local
Received: from mail-qg0-f54.google.com (mail-qg0-f54.google.com
 [209.85.192.54])
 (using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
 (Client CN "smtp.gmail.com",
 Issuer "Google Internet Authority G2" (not verified))
 by pixelated.wazokazi.is (Postfix) with ESMTPS id 4727320200C1
 for <me@pixelated.org>; Thu, 25 Feb 2016 21:17:13 +0100 (CET)
Received: by mail-qg0-f54.google.com with SMTP id b35so49276927qge.0
 for <me@pixelated.org>; Thu, 25 Feb 2016 12:17:12 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=thoughtworks-com.20150623.gappssmtp.com; s=20150623;
 h=content-type:subject:mime-version:from:date
 :content-transfer-encoding:message-id:content-description:to;
 bh=nveR36F1haWlVrLoKw9k9ZhuDgd3HqZEBMDfBsHbS+E=;
 b=EaYMcqBIABly0nSo+xTfWYAgaF+6V7bBE3HW9HTxB5Y7yvJLujn2+8o/zlDFmSsU5/
 E4Um9VVGJUkLYHhiS+9CQd4mTgqfOJmn0rRWk7mYWg4VOg47RQEUqI9gI+6am7vfHNzi
 ut48Wn5kJ8b65Sbx3wly+0xgZO+6Ti3w1Gv9Bo/Haiex4zBEW8pLY3GH2mluoN8I0kW3
 8W6tXtJ9tHuuh+s/lifRf/vCMQolIqA3azX31rYXkS9ePSDNE+ieRQOT7EO/aN3K6tJH
 usw1Ha2EkfTfnKpzpOpbafRnpwP0I23HIVWXxxWAWUfRLM8iJHvMYDKeGs9Ovypghm4H
 RGAg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=1e100.net; s=20130820;
 h=x-gm-message-state:content-type:subject:mime-version:from:date
 :content-transfer-encoding:message-id:content-description:to;
 bh=nveR36F1haWlVrLoKw9k9ZhuDgd3HqZEBMDfBsHbS+E=;
 b=RjqwU4nw/ZqnRoRHGnNKt6B4lqScfydpPTi92KdNhBKLfq6Xx+XpXJnOx2v+eUZPVz
 cpq7GZZXIQjS8SZbnkALTm8uer04qVAdBepLv8KNI2PixsYRhnNNJdKArS4madERpoBN
 seGeuLUTQxH4FWlg6uqqqZ81SFF1HfPcgEkd/zt7Olvhi3prASfvqb13U0kO77xHqEWD
 E5TOLItFFq1hbW6FZpe0UG7DURmIqzCNOa99Fd+HTnXftwvIKYXaAj3dum8V6whVAWT5
 2b/GYyh/+iryqkXs73V2TagMPiufS5tmIjWJz/2D7xMw7kKtiZBWW0Cls1ifogWHSko4
 h1zg==
X-Gm-Message-State: AG10YOTAHFMaoMWXw9R+S+fembYm50U5TG6IoNCnJmn5JUjeBw+wyK1tdKxP22hNlOn7gE9A
X-Received: by 10.140.232.15 with SMTP id d15mr62955439qhc.87.1456431431236;
 Thu, 25 Feb 2016 12:17:11 -0800 (PST)
Received: from [10.27.12.179] ([189.125.92.194])
 by smtp.gmail.com with ESMTPSA id 188sm3911768qhi.1.2016.02.25.12.17.09
 for <me@pixelated.org> (version=TLSv1/SSLv3 cipher=OTHER);
 Thu, 25 Feb 2016 12:17:10 -0800 (PST)
Subject: From Apple Mail
Mime-Version: 1.0 (Mac OS X Mail 8.2 \\(2104\\))
X-Pgp-Agent: GPGMail 2.5.2
From: Thais Siqueira <thais@domain.com>
Date: Thu, 25 Feb 2016 17:16:55 -0300
Content-Transfer-Encoding: 7bit
Message-Id: <309D14C0-1C36-4ABA-865D-B91BDBB85E87@thoughtworks.com>
Content-Description: OpenPGP encrypted message
To: me@pixelated.org
X-Mailer: Apple Mail (2.2104)
Content-Type: multipart/signed;
 boundary="Apple-Mail=_71B95569-2DFF-49EA-BDFB-FE1B4497223A";
 protocol="application/pgp-signature"; micalg=pgp-sha512
X-Leap-Encryption: decrypted

This is an OpenPGP/MIME encrypted message (RFC 2440 and 3156)
--Apple-Mail=_71B95569-2DFF-49EA-BDFB-FE1B4497223A
Content-Transfer-Encoding: 7bit
Content-Type: text/plain;
\tcharset=us-ascii

Email sent from apple mail.


--Apple-Mail=_71B95569-2DFF-49EA-BDFB-FE1B4497223A
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment;
\tfilename=signature.asc
Content-Type: application/pgp-signature;
\tname=signature.asc
Content-Description: Message signed with OpenPGP using GPGMail

-----BEGIN PGP SIGNATURE-----
Comment: GPGTools - https://gpgtools.org

iQIcBAEBCgAGBQJWz2FCAAoJENCM3F7anFCV7GQQAL87nctFBIAMuMo/aDITCzDg
y2D/boDQbozSVn+xtz9wzcCUzVlHnwVM8EAHpOf0WsHvkcU5hgiOpd3STIF5vxUX
gPTTZusOfoA1CGSeE1uFwf4eijDUuduhS5MqG4AQhuPL2cxYxOpeaitvwaU8fRsK
ItjO+KLF4vJDtQDeTBUUhtXT64Tl4sovLYiqnwud7fDV/3O6fsccIkxOva4G4OXY
RUWaGA1KUBwucISau/H1skwx0BWX9lGh9NPrk2GqKazH/S/63SWiU67UQ1Jx6G5M
0XJ/qQK8Vengx6P/h30ObozpMDJqgBobhLjMU2iLpEJ5Tv1mShlCNyw9LgA9kJad
zh1X8JCh+n3Ph8qVC0HfplrinmhChDI/DKjAtZRvJYFX8XmxhRZZBdSLNpPWxoZ/
guvBPqHzfFpOQ0tfMVhodf4yLElYXC4y+i4MiWEpid+TjiynlKOvgz2ApBSqHLDZ
Vid+dg9dpMaQkh977zo22raNlFKt6gA/z0qBjremUe4J8fY4W1A8RstLd6nrj4rl
y9UFbWOF9g1AerLSWLR/vr/ayKRFZDqvxZ4kkY5iKwDpbb9wxQ1/W/RmO2nBGqWN
TMUM64LxYb55980QFtxDdhc9okdgcchMVL6snhw7OgMynFswKQb9FZ/Gkpj38Sex
HJyPavR70Ldjh1vZMOSD
=bc/m
-----END PGP SIGNATURE-----

--Apple-Mail=_71B95569-2DFF-49EA-BDFB-FE1B4497223A--
```

** Thunderbird:

signkey: ```<OpenPGPKey 0x9DBC4ADD36277123 ([u'me2@mydomain.com', u'me@mydomain.com'] - publ)>```

decrdata:
```
From nobody Thu Feb 25 20:31:10 2016
Return-Path: <me@mydomain.com>
X-Original-To: me@pixelated.org
Delivered-To: 8823007cd15b6b2b72b15517da7630a9@deliver.local
Received: from mail-qg0-f45.google.com (mail-qg0-f45.google.com
 [209.85.192.45])
 (using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
 (Client CN "smtp.gmail.com",
 Issuer "Google Internet Authority G2" (not verified))
 by pixelated.wazokazi.is (Postfix) with ESMTPS id 84FC520200C1
 for <me@pixelated.org>; Thu, 25 Feb 2016 21:29:32 +0100 (CET)
Received: by mail-qg0-f45.google.com with SMTP id y89so49580661qge.2
 for <me@pixelated.org>; Thu, 25 Feb 2016 12:29:32 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=thoughtworks-com.20150623.gappssmtp.com; s=20150623;
 h=to:from:subject:message-id:date:user-agent:mime-version
 :content-type; bh=tMDtbO6gSMX/jHuYNBJbi9Y4K66CRPJprxo/1XQDnbs=;
 b=uQv3v4M5vLjMcuzwP4MzhRYD7pPS+VFk+kJbN2NFYSxK/wuza07GeZ0EpmyZ9pcvW4
 qrRj0Q33LWeOApM280VGC08Dp7WFm3xjUIgsc6MWOs1Q9+SXRwreG82SCVxZRYLHHgB0
 v/4vvtJ266o+LAX/gogBXUgLoLejzDRGLm2IWl5WIiBQ2Ploqmim/tvVelNtxQI5K2Ct
 SxMvmZqxQTGpa6M9HI7Ob85IXbxRs4y7nUwNNaAqxNH03YDQpT0+WkXhUaeStpqo2cXs
 VLakySn3Sn6ds1LVmbe+peOTDb5fCJcawCt4UJVywZ4TnBr9hKaWbJmPulmP4K2CQZcd
 XNkQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=1e100.net; s=20130820;
 h=x-gm-message-state:to:from:subject:message-id:date:user-agent
 :mime-version:content-type;
 bh=tMDtbO6gSMX/jHuYNBJbi9Y4K66CRPJprxo/1XQDnbs=;
 b=lZDvRoAhIoTtZ2x3GLAFoSmLOivwa+eyZlWm8TTERQ0bRTvIcY3x95rJ3CfcalfpjQ
 1GJxc8rRxFZ5IWqz16lU8Oh2xRo5ULI7s1OXmzIDtj8/Dn8GNaPDOaVPl6eolVv7yDH6
 mvNmhzX6Dz7vK8Eb+eioNwmPg7a9JFQKu1cuRGgiyMDzINp2PDTFfA+wGqd4QHs8fI3H
 4YtPHMLBogdU5KLHQityriY7UEYzZ0rWj7gf0Ym9fdavvKdoZVbk2qSzBaoCWgGm8Z01
 feC+hDLUO+xByIht1tqX409xhSConAgmvvOhAqabUYtsLbTINLIQLiyREGU31JguK7QZ
 PDow==
X-Gm-Message-State: AG10YORAZzXKVS5Z+ZOGKimLff9E9lXD+Jtc3guzkdf+eeXUNBbsIeDGbDIkLqZheeIGFD0x
X-Received: by 10.140.106.97 with SMTP id d88mr58578710qgf.84.1456432171190;
 Thu, 25 Feb 2016 12:29:31 -0800 (PST)
Received: from localhost.localdomain ([189.125.92.194])
 by smtp.gmail.com with ESMTPSA id t103sm3945370qgd.37.2016.02.25.12.29.29
 for <me@pixelated.org> (version=TLSv1/SSLv3 cipher=OTHER);
 Thu, 25 Feb 2016 12:29:30 -0800 (PST)
To: me@pixelated.org
From: Jefferson Stachelski <me@mydomain.com>
Subject: test
Message-ID: <56CF6427.1000901@thoughtworks.com>
Date: Thu, 25 Feb 2016 17:29:27 -0300
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:38.0) Gecko/20100101
 Thunderbird/38.5.0
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="ejC3daFl5FBtp7MoC9oUfPfXpbJwqlaqM"
X-Leap-Encryption: decrypted

This is an OpenPGP/MIME encrypted message (RFC 4880 and 3156)
--ejC3daFl5FBtp7MoC9oUfPfXpbJwqlaqM
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

signed and ecrypted email

--=20
Jefferson Stachelski | @jeffhsta
Security minded Developer
XMPP: jeffhsta@dukgo.com

--ejC3daFl5FBtp7MoC9oUfPfXpbJwqlaqM--
```

jeffhsta commented:
@kalikaneko the method _verify_signature_not_encrypted_msg returns a defer, but this method is added as a callback in another defer, it means that the next callback will get the result of the last callback on the defer returned by method _verify_signature_not_encrypted_msg.

So the line 469 that you mentioned have the value, not the defer, and the next callback "add_leap_header" will receive ret as a tuple.

About the try except is because sometimes is a mail with inline PGP signature, sometimes is not, so if it shows an error I just ignore assuming that there is no signature there.

Maybe replace this try, except for an if will be better, like:
if not isinstance(signkey, OpenPGPKey) and there_is_signature_to_extract(decrdata):
    return self._verify_signature_not_encrypted_msg(decrdata, senderAddress)
return (decrdata,signkey)

jeffhsta commented:
@kalikaneko I would like to refactor that code, I am not member of Pixelated more, and I need help from someone that have Apple Mail (I am running linux here).

Maybe @thaissiqueira can help on that, she was pairing with me on this, and she has OSX, so on her machine is possible reproduce that error

kalikaneko commented:
thanks @jeffhsta 
I see that @meskio touched some of this code, he might be able to help when he joins back next week too.

meskio commented:
I think the modification makes sense. I was worried about having this issue, I don't think Apple Mail is the onlye one that does `enc(sign(msg))`, both formats are part of the PGP/MIME standard.

I think the comments from @kalikaneko make sense and we can merge it as soon as they are fixed.

meskio commented:
About the try/except I'm surprised that `_verify_signature_not_encrypted_msg`, that will imply that `KeyManager.verify` raises an exception. I think they only return a failure the twisted way.

thaissiqueira commented:
@kalikaneko could you close this PR, please?
Because now we have the #243 that replace this one...

thaissiqueira commented:
@kalikaneko thanks :)
